tap_mailchimp/streams.py

```diff
@@ -0,0 +1,545 @@
+                formatted_field_names.append(self.data_key+'.'+field)
+
+        default_fields = ['_links', 'total_items', 'constraints', self.data_key+'.'+'_links']
+        formatted_field_names += default_fields
+        if self.extra_fields:
+            formatted_field_names += self.extra_fields
+        return ",".join(formatted_field_names)
+
+    def sync_substream(self, child, parent_record_id):
+        child_stream_obj = STREAMS.get(child)(self.state, self.client, self.config, self.catalog, self.selected_stream_names, self.child_streams_to_sync)
+        if child_stream_obj.replication_method == 'INCREMENTAL':
+            child_stream_obj.bookmark_path[1] = parent_record_id
+        child_stream_obj.path = self.get_path(parent_record_id, child_stream_obj)
+        child_stream_obj.sync()
+
+class FullTable(BaseStream):
+    replication_method = 'FULL_TABLE'
+
+    def sync(self):
+
+        ids = []
+        page_size = int(self.config.get('page_size', '1000'))
+        offset = 0
+        has_more = True
+        while has_more:
+            params = {
+                'count': page_size,
+                'offset': offset,
+                **self.params
+            }
+
+            LOGGER.info('%s - Syncing - count: %s, offset: %s', self.stream_name, page_size, offset)
+
+            formatted_selected_fields = self.format_selected_fields()
+
+            params['fields'] = formatted_selected_fields
+            data = self.client.get(
+                self.path,
+                params=params,
+                endpoint=self.stream_name)
+
+            raw_records = data.get(self.data_key)
+
+            if len(raw_records) < page_size:
+                has_more = False
+
+            for record in raw_records:
+                del record['_links']
+                ids.append(record.get('id'))
+                if self.child:
+                    for child in self.child:
+                        if child in self.report_streams:
+                            continue
+                        if child in self.selected_stream_names or child in self.child_streams_to_sync:
+                            self.sync_substream(child, record.get('id'))
+            if self.to_write_records:
+                self.process_records(raw_records)
+
+            for report_stream in self.report_streams:
+                if report_stream not in self.selected_stream_names:
+                    continue
+                reports_stream_obj = STREAMS.get(report_stream)(self.state, self.client, self.config, self.catalog, self.selected_stream_names, self.child_streams_to_sync)
+                reports_stream_obj.sync_report_activities(ids)
+
+            offset += page_size
+
+class Incremental(BaseStream):
+    replication_method = 'INCREMENTAL'
+
+    def sync(self):
+        last_datetime = self.get_bookmark(self.bookmark_path, self.config.get('start_date'))
```

Comment:
The sync() method of the incremental stream has almost the same code as the sync() method of the full_table stream except bookmarking logic. Please create one separate function like get_records() and move all the pagintaion-related code there, and use it in incremental and full_table sync.

Reply:
Updated the sync code for Full Table and Incremental streams as suggested.